Initial RV64 preparations

cerg2010cerg2010 opened the pull request.

cerg2010cerg2010 committed:
Initial RV64 preparations

Still lacks special RV64 instructions, however, current ones are
adapted (but not properly tested)

The following work is done:
- New types introduced: reg_t, sreg_t, virtaddr_t (same as reg_t, but
  it's semantics are a value holding virtual address),
  physaddr_t. Also special formatting identifiers (like PRIxaddr)
  are added.
- Move current instruction interpreter from hardcoded uint32_t type
  to reg_t/sreg_t. All instructions can now work with 64-bit and
  32-bit reg_t/sreg_t types.
- Replaced constants like 0xFFFFFFFD with more bitness-unaware
  constructs.
- 64-bit MMU implementation
- MXL/SXL/UXL fields support, as well as CSR width modulation upon ISA
  switch (not tested)
- Other minor changes

cerg2010cerg2010 committed:
Merge branch 'master' of https://github.com/LekKit/RVVM

cerg2010cerg2010 committed:
Merge branch 'master' of https://github.com/LekKit/RVVM

cerg2010cerg2010 committed:
Fix function types

LekKit commented on src/clint.c:
```diff
@@ -17,12 +17,13 @@ along with this program.  If not, see <https://www.gnu.org/licenses/>.
 #include "bit_ops.h"
 
-bool clint_mmio_handler(riscv32_vm_state_t* vm, riscv32_mmio_device_t* device, uint32_t offset, void* data, uint32_t size, uint8_t access)
+bool clint_mmio_handler(riscv32_vm_state_t* vm, riscv32_mmio_device_t* device, physaddr_t offset, void* data, uint32_t size, uint8_t access)
```
Change offset type to ptrdiff_t, since it is relative to MMIO device base address
Change size type to size_t
This applies to every MMIO callback

LekKit commented on src/riscv32.h:
```diff
@@ -66,9 +68,19 @@ enum
+
+#define ISA_MAX ISA_RV32
+
 #define INTERRUPT_MASK 0x80000000
```
This will be changed in future for proper RV64 support

LekKit commented on src/riscv32_csr_m.c:
```diff
@@ -17,30 +17,51 @@ You should have received a copy of the GNU General Public License
 
 #define CSR_MISA_RV32  0x40000000
 #define CSR_MISA_RV64  0x80000000
 #define CSR_MISA_RV128 0xC0000000
```
Needs to be shifted for RV64 as well

cerg2010cerg2010 replied:
It is, that's a removal. This constant is now replaced with csr_mstatus_mask function

LekKit commented on src/riscv32_csr_s.c:
```diff
@@ -17,84 +17,111 @@ You should have received a copy of the GNU General Public License
+    }
+    else
+    {
+	    satp = ((reg_t)vm->mmu_virtual << 60) | ((vm->root_page_table >> 12) & gen_mask(44));
```
GCC: warning: left shift count >= width of type
Probably should be separated with ifdef

LekKit commented on src/riscv32i.c:
```diff
@@ -119,16 +119,27 @@ static void riscv32i_srli_srai(riscv32_vm_state_t *vm, const uint32_t instructio
+		    src_reg &= gen_mask(extend_from);
+		    break;
+	    }
+    }
```
This should not be done here, separate instruction implementations with separate dispatch tables are going to be used for RV64. New srli/srai/slli are 3-4 times slower than the original implementation, and this is likely to be redundant. I suggest reverting this change.

LekKit commented on src/riscv32i.c:
```diff
@@ -119,16 +119,27 @@ static void riscv32i_srli_srai(riscv32_vm_state_t *vm, const uint32_t instructio
 
-    if (funct7 == 0x20) {
-        riscv32i_write_register_s(vm, rds, ((int32_t)src_reg) >> shamt);
+    if (funct7 == 1 << (12 - XLEN_BIT(vm) - 2)) {
```
GCC: warning: comparison of integer expressions of different signedness

LekKit commented on src/riscv32_mmu.c:
```diff
@@ -103,60 +148,144 @@ static void tlb_put(riscv32_vm_state_t* vm, uint32_t addr, uint32_t page_addr, u
 
 // Virtual memory addressing mode (SV32)
-static bool riscv32_mmu_translate_sv32(riscv32_vm_state_t* vm, uint32_t addr, uint8_t access, uint32_t* dest_addr)
+bool riscv_mmu_translate(riscv32_vm_state_t *vm, const virtaddr_t va, physaddr_t *pa, const uint8_t access, bool update_pages)
```
The soft MMU is a lot slower now, Linux system boot takes twice as much time, ramspeed reports similar regressions in raw memory speed. I think SV32 & SV39/48 translations should be separate functions as intended previously, and more optimally written, otherwise the compiler optimizations are less efficient.

LekKit commented on src/main.c:
```diff
@@ -111,60 +113,72 @@ int main(int argc, char** argv)
 	    }
 
-	    printf("DTB loaded at: 0x%zx size: %"PRId32"\n", dtb_addr, dts);
+	    printf("DTB loaded at: 0x%"PRIxpaddr" size: %"PRIxpaddr"\n", dtb_addr, dts);
```
Size is shown as hex without 0x prefix, not very straightforward - looks like incorrect decimal size

cerg2010cerg2010 committed:
Use old MMU translate function for SV32

This avoids performance loss of the new unified function. Hopefully
it will be optimized later.

cerg2010cerg2010 committed:
Introduce RV_64 macro, minor optimizations

This macro can be used to optimize certain paths which check current
XLEN mode. Other parts which check for ISA can be simplified later
in a similar way.

Also changed a couple of iterators in interpreter to char because
Clang's value range analysis fails on this kind of loops.
